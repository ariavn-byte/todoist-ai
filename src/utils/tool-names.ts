/**
 * Centralized tool names module
 *
 * This module provides a single source of truth for all tool names used throughout the codebase.
 * Each tool should import its own name from this module to avoid hardcoded strings.
 * This prevents outdated references when tool names change.
 */
export const ToolNames = {
    // Task management tools
    ADD_TASKS: 'add-tasks',
    COMPLETE_TASKS: 'complete-tasks',
    UPDATE_TASKS: 'update-tasks',
    FIND_TASKS: 'find-tasks',
    FIND_TASKS_BY_DATE: 'find-tasks-by-date',
    FIND_COMPLETED_TASKS: 'find-completed-tasks',

    // Project management tools
    ADD_PROJECTS: 'add-projects',
    UPDATE_PROJECTS: 'update-projects',
    FIND_PROJECTS: 'find-projects',

    // Section management tools
    ADD_SECTIONS: 'add-sections',
    UPDATE_SECTIONS: 'update-sections',
    FIND_SECTIONS: 'find-sections',

    // General tools
    GET_OVERVIEW: 'get-overview',
    DELETE_OBJECT: 'delete-object',
} as const

// Type for all tool names
export type ToolName = (typeof ToolNames)[keyof typeof ToolNames]
