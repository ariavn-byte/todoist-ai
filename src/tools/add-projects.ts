import type { PersonalProject, WorkspaceProject } from '@doist/todoist-api-typescript'
import { z } from 'zod'
import { getToolOutput } from '../mcp-helpers.js'
import type { TodoistTool } from '../todoist-tool.js'
import { formatNextSteps } from '../utils/response-builders.js'
import { ToolNames } from '../utils/tool-names.js'

const { ADD_SECTIONS, ADD_TASKS, FIND_PROJECTS, GET_OVERVIEW } = ToolNames

const ProjectSchema = z.object({
    name: z.string().min(1).describe('The name of the project.'),
    isFavorite: z
        .boolean()
        .optional()
        .describe('Whether the project is a favorite. Defaults to false.'),
    viewStyle: z
        .enum(['list', 'board', 'calendar'])
        .optional()
        .describe('The project view style. Defaults to "list".'),
})

const ArgsSchema = {
    projects: z.array(ProjectSchema).min(1).describe('The array of projects to add.'),
}

const addProjects = {
    name: ToolNames.ADD_PROJECTS,
    description: 'Add one or more new projects.',
    parameters: ArgsSchema,
    async execute({ projects }, client) {
        const newProjects = await Promise.all(projects.map((project) => client.addProject(project)))
        const textContent = generateTextContent({ projects: newProjects })

        return getToolOutput({
            textContent,
            structuredContent: {
                projects: newProjects,
                totalCount: newProjects.length,
            },
        })
    },
} satisfies TodoistTool<typeof ArgsSchema>

function generateTextContent({
    projects,
}: {
    projects: (PersonalProject | WorkspaceProject)[]
}) {
    const count = projects.length
    const projectList = projects.map((project) => `â€¢ ${project.name} (id=${project.id})`).join('\n')

    const summary = `Added ${count} project${count === 1 ? '' : 's'}:\n${projectList}`

    // Context-aware next steps for new projects
    const nextSteps: string[] = []

    if (count === 1) {
        const project = projects[0]
        if (project) {
            nextSteps.push(`Use ${ADD_SECTIONS} to organize new project with sections`)
            nextSteps.push(`Use ${ADD_TASKS} to add your first tasks to this project`)
            nextSteps.push(
                `Use ${GET_OVERVIEW} with projectId=${project.id} to see project structure`,
            )
        }
    } else {
        nextSteps.push(`Use ${ADD_SECTIONS} to organize these projects with sections`)
        nextSteps.push(`Use ${ADD_TASKS} to add tasks to these projects`)
        nextSteps.push(`Use ${FIND_PROJECTS} to see all projects including the new ones`)
        nextSteps.push(`Use ${GET_OVERVIEW} to see updated project hierarchy`)
    }

    const next = formatNextSteps(nextSteps)
    return `${summary}\n${next}`
}

export { addProjects }
