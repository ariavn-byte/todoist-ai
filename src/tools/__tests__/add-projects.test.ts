import type { PersonalProject, TodoistApi, WorkspaceProject } from '@doist/todoist-api-typescript'
import { jest } from '@jest/globals'
import {
    TEST_IDS,
    createMockProject,
    extractStructuredContent,
    extractTextContent,
} from '../../utils/test-helpers.js'
import { ToolNames } from '../../utils/tool-names.js'
import { addProjects } from '../add-projects.js'

// Mock the Todoist API
const mockTodoistApi = {
    addProject: jest.fn(),
} as unknown as jest.Mocked<TodoistApi>

const { ADD_TASKS, ADD_PROJECTS, ADD_SECTIONS } = ToolNames

describe(`${ADD_PROJECTS} tool`, () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })

    describe('creating a single project', () => {
        it('should create a project and return mapped result', async () => {
            const mockApiResponse = createMockProject({
                id: TEST_IDS.PROJECT_TEST,
                name: 'test-abc123def456-project',
                childOrder: 1,
                createdAt: '2024-01-01T00:00:00Z',
            })

            mockTodoistApi.addProject.mockResolvedValue(mockApiResponse)

            const result = await addProjects.execute(
                { projects: [{ name: 'test-abc123def456-project' }] },
                mockTodoistApi,
            )

            // Verify API was called correctly
            expect(mockTodoistApi.addProject).toHaveBeenCalledWith({
                name: 'test-abc123def456-project',
            })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Added 1 project:')
            expect(textContent).toContain('test-abc123def456-project')
            expect(textContent).toContain(`id=${TEST_IDS.PROJECT_TEST}`)
            expect(textContent).toContain(`Use ${ADD_TASKS} to add your first tasks`)

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    projects: [
                        expect.objectContaining({
                            id: TEST_IDS.PROJECT_TEST,
                            name: 'test-abc123def456-project',
                        }),
                    ],
                    totalCount: 1,
                }),
            )
        })

        it('should handle different project properties from API', async () => {
            const mockApiResponse = createMockProject({
                id: 'project-456',
                name: 'My Blue Project',
                color: 'blue',
                isFavorite: true,
                isShared: true,
                parentId: 'parent-123',
                viewStyle: 'board',
                childOrder: 2,
                description: 'A test project',
                createdAt: '2024-01-01T00:00:00Z',
            })

            mockTodoistApi.addProject.mockResolvedValue(mockApiResponse)

            const result = await addProjects.execute(
                { projects: [{ name: 'My Blue Project' }] },
                mockTodoistApi,
            )

            expect(mockTodoistApi.addProject).toHaveBeenCalledWith({
                name: 'My Blue Project',
                isFavorite: undefined,
                viewStyle: undefined,
            })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Added 1 project:')
            expect(textContent).toContain('My Blue Project')
            expect(textContent).toContain('id=project-456')
            expect(textContent).toContain(`Use ${ADD_SECTIONS} to organize new project`)
        })

        it('should create project with isFavorite and viewStyle options', async () => {
            const mockApiResponse = createMockProject({
                id: 'project-789',
                name: 'Board Project',
                isFavorite: true,
                viewStyle: 'board',
            })

            mockTodoistApi.addProject.mockResolvedValue(mockApiResponse)

            const result = await addProjects.execute(
                { projects: [{ name: 'Board Project', isFavorite: true, viewStyle: 'board' }] },
                mockTodoistApi,
            )

            expect(mockTodoistApi.addProject).toHaveBeenCalledWith({
                name: 'Board Project',
                isFavorite: true,
                viewStyle: 'board',
            })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Added 1 project:')
            expect(textContent).toContain('Board Project')
            expect(textContent).toContain('id=project-789')
        })
    })

    describe('creating multiple projects', () => {
        it('should create multiple projects and return mapped results', async () => {
            type Project = PersonalProject | WorkspaceProject
            const mockProjects: [Project, Project, Project] = [
                createMockProject({ id: 'project-1', name: 'First Project' }),
                createMockProject({ id: 'project-2', name: 'Second Project' }),
                createMockProject({ id: 'project-3', name: 'Third Project' }),
            ]

            const [project1, project2, project3] = mockProjects
            mockTodoistApi.addProject
                .mockResolvedValueOnce(project1)
                .mockResolvedValueOnce(project2)
                .mockResolvedValueOnce(project3)

            const result = await addProjects.execute(
                {
                    projects: [
                        { name: 'First Project' },
                        { name: 'Second Project' },
                        { name: 'Third Project' },
                    ],
                },
                mockTodoistApi,
            )

            // Verify API was called correctly for each project
            expect(mockTodoistApi.addProject).toHaveBeenCalledTimes(3)
            expect(mockTodoistApi.addProject).toHaveBeenNthCalledWith(1, { name: 'First Project' })
            expect(mockTodoistApi.addProject).toHaveBeenNthCalledWith(2, { name: 'Second Project' })
            expect(mockTodoistApi.addProject).toHaveBeenNthCalledWith(3, { name: 'Third Project' })

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Added 3 projects:')
            expect(textContent).toContain('First Project (id=project-1)')
            expect(textContent).toContain('Second Project (id=project-2)')
            expect(textContent).toContain('Third Project (id=project-3)')
            expect(textContent).toContain(`Use ${ADD_SECTIONS} to organize these projects`)

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    projects: mockProjects,
                    totalCount: 3,
                }),
            )
        })
    })

    describe('error handling', () => {
        it('should propagate API errors', async () => {
            const apiError = new Error('API Error: Project name is required')
            mockTodoistApi.addProject.mockRejectedValue(apiError)

            await expect(
                addProjects.execute({ projects: [{ name: '' }] }, mockTodoistApi),
            ).rejects.toThrow('API Error: Project name is required')
        })

        it('should handle partial failures in multiple projects', async () => {
            const mockProject = createMockProject({
                id: 'project-1',
                name: 'First Project',
            })

            mockTodoistApi.addProject
                .mockResolvedValueOnce(mockProject)
                .mockRejectedValueOnce(new Error('API Error: Invalid project name'))

            await expect(
                addProjects.execute(
                    {
                        projects: [{ name: 'First Project' }, { name: 'Invalid' }],
                    },
                    mockTodoistApi,
                ),
            ).rejects.toThrow('API Error: Invalid project name')
        })
    })
})
