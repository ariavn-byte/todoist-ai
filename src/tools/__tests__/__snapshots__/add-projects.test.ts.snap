// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`add-projects tool creating a single project should create a project and return mapped result 1`] = `
"Added 1 project:
• test-abc123def456-project (id=6cfCcrrCFg2xP94Q)
Next:
- Use add-sections to organize new project with sections
- Use add-tasks to add your first tasks to this project
- Use get-overview with projectId=6cfCcrrCFg2xP94Q to see project structure"
`;

exports[`add-projects tool creating a single project should create project with isFavorite and viewStyle options 1`] = `
"Added 1 project:
• Board Project (id=project-789)
Next:
- Use add-sections to organize new project with sections
- Use add-tasks to add your first tasks to this project
- Use get-overview with projectId=project-789 to see project structure"
`;

exports[`add-projects tool creating a single project should handle different project properties from API 1`] = `
"Added 1 project:
• My Blue Project (id=project-456)
Next:
- Use add-sections to organize new project with sections
- Use add-tasks to add your first tasks to this project
- Use get-overview with projectId=project-456 to see project structure"
`;

exports[`add-projects tool creating multiple projects should create multiple projects and return mapped results 1`] = `
"Added 3 projects:
• First Project (id=project-1)
• Second Project (id=project-2)
• Third Project (id=project-3)
Next:
- Use add-sections to organize these projects with sections
- Use add-tasks to add tasks to these projects
- Use find-projects to see all projects including the new ones
- Use get-overview to see updated project hierarchy"
`;
